File: src/values/bool.rs
use crate::bytecodes::ApicaTypeBytecode;
use crate::values::_type::ValueType;
use crate::values::char::ValueChar;
use crate::values::f32::ValueF32;
use crate::values::f64::ValueF64;
use crate::values::i16::ValueI16;
use crate::values::i32::ValueI32;
use crate::values::i64::ValueI64;
use crate::values::i8::ValueI8;
use crate::values::string::ValueString;
use crate::values::u16::ValueU16;
use crate::values::u32::ValueU32;
use crate::values::u64::ValueU64;
use crate::values::u8::ValueU8;
use crate::values::value::Value;

#[derive(Clone)]
pub struct ValueBool {
    value: Option<bool>,
}

impl ValueBool {
    pub fn init_empty() -> ValueBool {
        ValueBool { value: None }
    }
    
    pub fn init_with(value: bool) -> ValueBool {
        ValueBool { value: Some(value) }
    }
    
    pub fn show(&self, end: char) {
        if let Some(val) = self.value {
            print!("{val}{end}");
        } else {
            print!("null{end}");
        }
    }
    
    pub fn is_null(&self) -> bool {
        self.value.is_none()
    }
    
    pub fn get_type_representation(&self) -> &str {
        "bool"
    }
    
    pub fn get_value(&self) -> Option<bool> {
        self.value
    }

    pub fn add(&self, other: &Value) -> Option<Value> {
        match other {
            Value::I8(i8) => Some(Value::I8(ValueI8::init_with(if self.value.unwrap() { 1 } else { 0 } + i8.get_value().unwrap()))),
            Value::I16(i16) => Some(Value::I16(ValueI16::init_with(if self.value.unwrap() { 1 } else { 0 } + i16.get_value().unwrap()))),
            Value::I32(i32) => Some(Value::I32(ValueI32::init_with(if self.value.unwrap() { 1 } else { 0 } + i32.get_value().unwrap()))),
            Value::I64(i64) => Some(Value::I64(ValueI64::init_with(if self.value.unwrap() { 1 } else { 0 } + i64.get_value().unwrap()))),
            Value::U8(u8) => Some(Value::U8(ValueU8::init_with(if self.value.unwrap() { 1 } else { 0 } + u8.get_value().unwrap()))),
            Value::U16(u16) => Some(Value::U16(ValueU16::init_with(if self.value.unwrap() { 1 } else { 0 } + u16.get_value().unwrap()))),
            Value::U32(u32) => Some(Value::U32(ValueU32::init_with(if self.value.unwrap() { 1 } else { 0 } + u32.get_value().unwrap()))),
            Value::U64(u64) => Some(Value::U64(ValueU64::init_with(if self.value.unwrap() { 1 } else { 0 } + u64.get_value().unwrap()))),

            Value::F32(f32) => Some(Value::F32(ValueF32::init_with(if self.value.unwrap() { 1.0 } else { 0.0 } + f32.get_value().unwrap()))),
            Value::F64(f64) => Some(Value::F64(ValueF64::init_with(if self.value.unwrap() { 1.0 } else { 0.0 } + f64.get_value().unwrap()))),
            Value::Bool(bool) => Some(Value::U8(ValueU8::init_with(if self.value.unwrap() { 1 } else { 0 } + if bool.get_value().unwrap() { 1 } else { 0 }))),

            Value::Char(char) => Some(Value::U32(ValueU32::init_with(if self.value.unwrap() { 1 } else { 0 } + char.get_value().unwrap() as u32))),

            _ => None,
        }
    }
    
    pub fn not(&self) -> Value {
        let value = match self.value {
            Some(value) => !value,
            None => false,
        };
        Value::Bool(ValueBool::init_with(value))
    }
    
    pub fn convert(&self, to: ApicaTypeBytecode) -> Option<Value> {
        if let Some(value) = &self.value {
            match to {
                ApicaTypeBytecode::Char => Some(Value::Char(ValueChar::init_with(*value as u8 as char))),
                ApicaTypeBytecode::String => Some(Value::String(ValueString::init_with(format!("{value}")))),
                ApicaTypeBytecode::Type => Some(Value::Type(ValueType::init_with(ApicaTypeBytecode::Bool, None))),

                _ => None,
            }
        } else {
            match to {
                ApicaTypeBytecode::Char => Some(Value::Char(ValueChar::init_empty())),
                ApicaTypeBytecode::String => Some(Value::String(ValueString::init_empty())),
                ApicaTypeBytecode::Type => Some(Value::Type(ValueType::init_with(ApicaTypeBytecode::Bool, None))),

                _ => None,
            }
        }
    }

    pub fn auto_convert(&self, to: ApicaTypeBytecode) -> Option<Value> {
        if let Some(value) = &self.value {
            match to {
                ApicaTypeBytecode::Any => Some(Value::Bool(ValueBool::init_with(*value))),
                
                ApicaTypeBytecode::I8 => Some(Value::I8(ValueI8::init_with(*value as i8))),
                ApicaTypeBytecode::I16 => Some(Value::I16(ValueI16::init_with(*value as i16))),
                ApicaTypeBytecode::I32 => Some(Value::I32(ValueI32::init_with(*value as i32))),
                ApicaTypeBytecode::I64 => Some(Value::I64(ValueI64::init_with(*value as i64))),
                ApicaTypeBytecode::U8 => Some(Value::U8(ValueU8::init_with(*value as u8))),
                ApicaTypeBytecode::U16 => Some(Value::U16(ValueU16::init_with(*value as u16))),
                ApicaTypeBytecode::U32 => Some(Value::U32(ValueU32::init_with(*value as u32))),
                ApicaTypeBytecode::U64 => Some(Value::U64(ValueU64::init_with(*value as u64))),
                ApicaTypeBytecode::F32 => Some(Value::F32(ValueF32::init_with(if *value { 1.0 } else { 0.0 }))),
                ApicaTypeBytecode::F64 => Some(Value::F64(ValueF64::init_with(if *value { 1.0 } else { 0.0 }))),
                ApicaTypeBytecode::Bool => Some(Value::Bool(ValueBool::init_with(*value))),

                _ => None,
            }
        } else {
            match to {
                ApicaTypeBytecode::Any => Some(Value::Bool(ValueBool::init_empty())),
                
                ApicaTypeBytecode::I8 => Some(Value::I8(ValueI8::init_empty())),
                ApicaTypeBytecode::I16 => Some(Value::I16(ValueI16::init_empty())),
                ApicaTypeBytecode::I32 => Some(Value::I32(ValueI32::init_empty())),
                ApicaTypeBytecode::I64 => Some(Value::I64(ValueI64::init_empty())),
                ApicaTypeBytecode::U8 => Some(Value::U8(ValueU8::init_empty())),
                ApicaTypeBytecode::U16 => Some(Value::U16(ValueU16::init_empty())),
                ApicaTypeBytecode::U32 => Some(Value::U32(ValueU32::init_empty())),
                ApicaTypeBytecode::U64 => Some(Value::U64(ValueU64::init_empty())),
                ApicaTypeBytecode::F32 => Some(Value::F32(ValueF32::init_empty())),
                ApicaTypeBytecode::F64 => Some(Value::F64(ValueF64::init_empty())),
                ApicaTypeBytecode::Bool => Some(Value::Bool(ValueBool::init_empty())),

                _ => None,
            }
        }
    }
}

File: src/values/bool.cpp
#include "values/bool.hpp"
#include "values/i8.hpp"
#include "values/i16.hpp"
#include "values/i32.hpp"
#include "values/i64.hpp"
#include "values/u8.hpp"
#include "values/u16.hpp"
#include "values/u32.hpp"
#include "values/u64.hpp"
#include "values/f32.hpp"
#include "values/f64.hpp"
#include "values/char.hpp"
#include "values/string.hpp"
#include "values/type.hpp"
#include "elements.hpp"

using namespace common::values;

ValueBool::ValueBool()
    : value(std::nullopt) {

}

ValueBool::ValueBool(bool value)
    : value(value) {
    
}

void ValueBool::show(char end) const {
    if (this->value)
        std::cout << "bool<" << (this->value.value() ? "true" : "false") << '>' << end;
    else
        std::cout << "bool<null>" << end;
}

bool ValueBool::isNull() const {
    return !this->value.has_value();
}

std::string ValueBool::getTypeRepr() const {
    return "bool";
}

common::bytecodes::ApicaTypeBytecode ValueBool::getKind() const {
    return common::bytecodes::ApicaTypeBytecode::Bool;
}

std::optional<Value*> ValueBool::add(const Value *other) const {
    switch (other->getKind()) {
        case common::bytecodes::ApicaTypeBytecode::I8: {
            const ValueI8 *i8 = static_cast<const ValueI8*>(other);
            return new ValueI8(i8->getValue().value() + this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::I16: {
            const ValueI16 *i16 = static_cast<const ValueI16*>(other);
            return new ValueI16(i16->getValue().value() + this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::I32: {
            const ValueI32 *i32 = static_cast<const ValueI32*>(other);
            return new ValueI32(i32->getValue().value() + this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::I64: {
            const ValueI64 *i64 = static_cast<const ValueI64*>(other);
            return new ValueI64(i64->getValue().value() + this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::U8: {
            const ValueU8 *u8 = static_cast<const ValueU8*>(other);
            return new ValueU8(u8->getValue().value() + this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::U16: {
            const ValueU16 *u16 = static_cast<const ValueU16*>(other);
            return new ValueU16(u16->getValue().value() + this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::U32: {
            const ValueU32 *u32 = static_cast<const ValueU32*>(other);
            return new ValueU32(u32->getValue().value() + this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::U64: {
            const ValueU64 *u64 = static_cast<const ValueU64*>(other);
            return new ValueU64(u64->getValue().value() + this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::F32: {
            const ValueF32 *f32 = static_cast<const ValueF32*>(other);
            return new ValueF32(f32->getValue().value() + this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::F64: {
            const ValueF64 *f64 = static_cast<const ValueF64*>(other);
            return new ValueF64(f64->getValue().value() + this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::Bool: {
            const ValueBool *boolean = static_cast<const ValueBool*>(other);
            return new ValueU8(boolean->getValue().value() + this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::Char: {
            const ValueChar *character = static_cast<const ValueChar*>(other);
            return new ValueU32(character->getValue().value() + this->value.value());
        }

        default: return std::nullopt;
    }
}

std::optional<Value*> ValueBool::increment() {
    return std::nullopt;
}

std::optional<Value*> ValueBool::subtract(const Value *other) const {
    switch (other->getKind()) {
        case common::bytecodes::ApicaTypeBytecode::I8: {
            const ValueI8 *i8 = static_cast<const ValueI8*>(other);
            return new ValueI8(i8->getValue().value() - this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::I16: {
            const ValueI16 *i16 = static_cast<const ValueI16*>(other);
            return new ValueI16(i16->getValue().value() - this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::I32: {
            const ValueI32 *i32 = static_cast<const ValueI32*>(other);
            return new ValueI32(i32->getValue().value() - this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::I64: {
            const ValueI64 *i64 = static_cast<const ValueI64*>(other);
            return new ValueI64(i64->getValue().value() - this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::U8: {
            const ValueU8 *u8 = static_cast<const ValueU8*>(other);
            return new ValueU8(u8->getValue().value() - this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::U16: {
            const ValueU16 *u16 = static_cast<const ValueU16*>(other);
            return new ValueU16(u16->getValue().value() - this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::U32: {
            const ValueU32 *u32 = static_cast<const ValueU32*>(other);
            return new ValueU32(u32->getValue().value() - this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::U64: {
            const ValueU64 *u64 = static_cast<const ValueU64*>(other);
            return new ValueU64(u64->getValue().value() - this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::F32: {
            const ValueF32 *f32 = static_cast<const ValueF32*>(other);
            return new ValueF32(f32->getValue().value() - this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::F64: {
            const ValueF64 *f64 = static_cast<const ValueF64*>(other);
            return new ValueF64(f64->getValue().value() - this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::Bool: {
            const ValueBool *boolean = static_cast<const ValueBool*>(other);
            return new ValueU8(boolean->getValue().value() - this->value.value());
        }

        case common::bytecodes::ApicaTypeBytecode::Char: {
            const ValueChar *character = static_cast<const ValueChar*>(other);
            return new ValueU32(character->getValue().value() - this->value.value());
        }

        default: return std::nullopt;
    }
}
std::optional<Value*> ValueBool::decrement() {
    return std::nullopt;
}

std::optional<Value*> ValueBool::unaryNot() const {
    return new ValueBool(this->value.has_value() ? !this->value.value() : true);
}

std::optional<Value*> ValueBool::convert(common::bytecodes::ApicaTypeBytecode to) const {
    if (this->value) {
        switch (to) {
            case common::bytecodes::ApicaTypeBytecode::Char:
                return new ValueChar(this->value.value());
            
            case common::bytecodes::ApicaTypeBytecode::String:
                return new ValueString(this->value.value() ? "true" : "false");
            
            case common::bytecodes::ApicaTypeBytecode::Type:
                return new ValueType(common::bytecodes::ApicaTypeBytecode::Bool);
            
            default: return std::nullopt;
        }
    } else {
        switch (to) {
            case common::bytecodes::ApicaTypeBytecode::Char:
                return new ValueChar();
            
            case common::bytecodes::ApicaTypeBytecode::String:
                return new ValueString();
            
            case common::bytecodes::ApicaTypeBytecode::Type:
                return new ValueType(common::bytecodes::ApicaTypeBytecode::Bool);
            
            default: return std::nullopt;
        }
    }
}

std::optional<Value*> ValueBool::autoConvert(common::bytecodes::ApicaTypeBytecode to) const {
    if (this->value) {
        switch (to) {
            case common::bytecodes::ApicaTypeBytecode::Any:
            case common::bytecodes::ApicaTypeBytecode::Bool:
                return new ValueBool(this->value.value());
            
            case common::bytecodes::ApicaTypeBytecode::I8:
                return new ValueI8(this->value.value());
            
            case common::bytecodes::ApicaTypeBytecode::I16:
                return new ValueI16(this->value.value());
            
            case common::bytecodes::ApicaTypeBytecode::I32:
                return new ValueI32(this->value.value());
            
            case common::bytecodes::ApicaTypeBytecode::I64:
                return new ValueI64(this->value.value());
            
            case common::bytecodes::ApicaTypeBytecode::U8:
                return new ValueU8(this->value.value());
            
            case common::bytecodes::ApicaTypeBytecode::U16:
                return new ValueU16(this->value.value());
            
            case common::bytecodes::ApicaTypeBytecode::U32:
                return new ValueU32(this->value.value());
            
            case common::bytecodes::ApicaTypeBytecode::U64:
                return new ValueU64(this->value.value());
            
            case common::bytecodes::ApicaTypeBytecode::F32:
                return new ValueF32(this->value.value());
            
            case common::bytecodes::ApicaTypeBytecode::F64:
                return new ValueF64(this->value.value());

            default: return std::nullopt;
        }
    } else {
        switch (to) {
            case common::bytecodes::ApicaTypeBytecode::Any:
            case common::bytecodes::ApicaTypeBytecode::Bool:
                return new ValueBool();
            
            case common::bytecodes::ApicaTypeBytecode::I8:
                return new ValueI8();
            
            case common::bytecodes::ApicaTypeBytecode::I16:
                return new ValueI16();
            
            case common::bytecodes::ApicaTypeBytecode::I32:
                return new ValueI32();
            
            case common::bytecodes::ApicaTypeBytecode::I64:
                return new ValueI64();
            
            case common::bytecodes::ApicaTypeBytecode::U8:
                return new ValueU8();
            
            case common::bytecodes::ApicaTypeBytecode::U16:
                return new ValueU16();
            
            case common::bytecodes::ApicaTypeBytecode::U32:
                return new ValueU32();
            
            case common::bytecodes::ApicaTypeBytecode::U64:
                return new ValueU64();
            
            case common::bytecodes::ApicaTypeBytecode::F32:
                return new ValueF32();
            
            case common::bytecodes::ApicaTypeBytecode::F64:
                return new ValueF64();

            default: return std::nullopt;
        }
    }
}

std::optional<bool> ValueBool::getValue() const {
    return this->value;
}